internal/datanode/writebuffer/insert_buffer.go

```diff
@@ -87,40 +97,96 @@ func NewInsertBuffer(sch *schemapb.CollectionSchema) (*InsertBuffer, error) {
 	if estSize == 0 {
 		return nil, errors.New("Invalid schema")
 	}
-	buffer, err := storage.NewInsertData(sch)
-	if err != nil {
-		return nil, err
-	}
+
 	sizeLimit := paramtable.Get().DataNodeCfg.FlushInsertBufferSize.GetAsInt64()
+	chunkSize := paramtable.Get().DataNodeCfg.InsertBufferChunkSize.GetAsInt64()
+	// use size Limit when chunkSize not valid
+	if chunkSize <= 0 || chunkSize > sizeLimit {
+		log.Warn("invalidate chunk size, use insert buffer size", zap.Int64("chunkSize", chunkSize), zap.Int64("insertBufferSize", sizeLimit))
+		chunkSize = sizeLimit
+	}
 
-	return &InsertBuffer{
+	estRow := chunkSize / int64(estSize)
+	ib := &InsertBuffer{
 		BufferBase: BufferBase{
 			rowLimit:      noLimit,
 			sizeLimit:     sizeLimit,
 			TimestampFrom: math.MaxUint64,
 			TimestampTo:   0,
 		},
 		collSchema: sch,
-		buffer:     buffer,
-	}, nil
+		estRow:     estRow,
+		chunkSize:  chunkSize,
+		buffers:    make([]*InsertBufferChunk, 0, sizeLimit/chunkSize),
+	}
+	err = ib.nextBatch()
+	if err != nil {
+		return nil, err
+	}
+
+	return ib, nil
+}
+
+func (ib *InsertBuffer) nextBatch() error {
+	buffer, err := storage.NewInsertDataWithCap(ib.collSchema, int(ib.estRow))
+	if err != nil {
+		return err
+	}
+	ib.buffers = append(ib.buffers, &InsertBufferChunk{
+		BufferBase: BufferBase{
+			rowLimit:      ib.estRow,
+			sizeLimit:     ib.chunkSize,
+			TimestampFrom: math.MaxUint64,
+			TimestampTo:   0,
+		},
+		buffer: buffer,
+	})
+	return nil
+}
+
+func (ib *InsertBuffer) currentBuffer() *InsertBufferChunk {
+	idx := len(ib.buffers) - 1
+	if idx < 0 || ib.buffers[idx].IsFull() {
+		ib.nextBatch()
+		idx++
+	}
+	return ib.buffers[idx]
+}
+
+func (ib *InsertBuffer) buffer(inData *storage.InsertData, tr TimeRange, startPos, endPos *msgpb.MsgPosition) {
+	buffer := ib.currentBuffer()
+	storage.MergeInsertData(buffer.buffer, inData)
```

Comment:
Can we just make the `inData` a single buffer to avoid this merge? What is the benefit if we keep a list of equally sized `BufferChunk`?

Reply:
> Can we just make the `inData` a single buffer to avoid this merge? What is the benefit if we keep a list of equally sized `BufferChunk`?

save some memory if we have many collections/partitions?